### ISSUE TITLE
给 oneflow.F 下的 functional API 添加 docstring 推进表

### ISSUE THREAD

**Issue by doombeaker** [Oneflow-Inc/oneflow#5545]
目前 `oneflow.F` 下有 120+个 functional API 没有 docstring。
可以参考 https://github.com/Oneflow-Inc/oneflow/pull/5449/files 为它们添加 docstring。

任务难度：低
需要做的事情：参考 `functional_api.yaml` 文件中的函数原型（signature）以及代码，了解 API 的用法，再添加 docstring（含doctest）。

本主贴记录周计划，以下 comment 有完整计划，这项工作欢迎开源社区的人员参与。

**备注**：
1. 期待参与者用 fork 的方式提交代码
2. 后期会组织针对开源社区的简单交流，科普如何通过 github 协作（fork、提交 PR、OneFlow 的 CI 流程），可关注 QQ 群 331883

## 周计划（7月27~7月30）
| functional API            | 认领人 | reviewer | PR    | 备注 | 放置到的文件            |
| ------------------------- | ------ | -------- | ----- | ---- | --------------- |
| F.abs                     | 张舸航 |          | #5632 |      | math_ops.py |
| F.ceil                    | 张长昊 |          | #5638       |      | math_ops.py     |
| F.diag                    | 谢子鹏 |          |  #5645      |      | array_ops.py    |
| F.exp                     | 党凯   |          |  #5624     |      | math_ops.py     |
| F.floor                   | 刘圻   |          |   https://github.com/Oneflow-Inc/oneflow/pull/5671    |      | math_ops.py     |
| F.acos                    | 张舸航 |          |  #5632   |      | math_ops.py     |
| F.acosh                   | 谢子鹏 |          |  #5645     |      | math_ops.py     |
| F.add                     | 谢子鹏 |          |  #5645     |      | math_ops.py     |
| F.add_n                   | 谢子鹏 |          |  #5645      |      | math_ops.py     |
| F.add_scalar              | 谢子鹏 |          | #5645      |      | math_ops.py     |
| F.add_scalar_by_tensor    | 张长昊 |           | #5638        |解释的可能有点问题      | math_ops.py     |
| F.argmax                  | 谢子鹏 |         | #5645      |      | math_ops.py     |
| F.argwhere                | 谢子鹏 |          |  #5645     |      | array_ops.py    |
| F.asin                    | 谢子鹏 |          |  #5645     |      | math_ops.py     |
| F.asinh                   | 谢子鹏 |          | #5645      |      | math_ops.py     |
| F.atan                    | 谢子鹏 |          |  #5645     |      | math_ops.py     |
| F.atanh                   | 张舸航 |          |   #5632   |      | math_ops.py     |
| F.avg_pool_2d             | 党凯   |          |  #5624  | 待定，和pytorch不对其     | array_ops.py    |
| F.batch_matmul            | 党凯   |          |  #5624      |  待定，部分参数和torch.bmm不对其    | math_ops.py     |
| F.bias_add                | 党凯   |          |  #5624     |      | math_ops.py     |
| F.broadcast_add           | 张长昊 |          | #5638       |      | math_ops.py     |
| F.broadcast_div           | 张长昊 |          | #5638       |      | math_ops.py     |
| F.broadcast_equal         | 党凯   |          | #5624   |      | math_ops.py     |
| F.broadcast_greater       | 张长昊 |          | #5638        |      | math_ops.py     |
| F.broadcast_greater_equal | 党凯   |          | #5624      |      | math_ops.py     |
| F.broadcast_less          | 张舸航 |          |  #5632     |      | math_ops.py     |
| F.broadcast_less_equal    | 张舸航 |          |   #5632    |      | math_ops.py     |
| F.broadcast_like          | 刘圻 |    |    https://github.com/Oneflow-Inc/oneflow/pull/5671   |   和主流框架不同，在解决   | array_ops.py    |
| F.broadcast_matmul        | 党凯   |          | #5624      |      | math_ops.py     |
| F.broadcast_mul           | 党凯   |          | #5624   |      | math_ops.py     |
| F.broadcast_not_equal     | 刘圻 |          |    https://github.com/Oneflow-Inc/oneflow/pull/5671   |      | math_ops.py     |
| F.broadcast_sub           | 刘圻 |          |  https://github.com/Oneflow-Inc/oneflow/pull/5671     |      | math_ops.py     |
| F.cast                    | 党凯   |          | #5624      |      | tensor_type.py                |
| F.clip_by_scalar          | 党凯   |          | #5624      |      | array_ops.py    |
| F.clip_by_scalar_max      | 党凯   |          | #5624      |      | array_ops.py    |
| F.clip_by_scalar_min      | 党凯   |          | #5624      |      | array_ops.py    |
| F.log                  |  张舸航      |          |  #5632    |      | math_ops.py |
| F.log1p                |  张舸航      |          |   #5632   |      | math_ops.py |
| F.pow                  |  张舸航      |        | #5632 |存在支持类型问题，已解决     | math_ops.py|
| F.sqrt                 |  张舸航      |        |   #5632       |      | math_ops.py|
|F.square |  刘圻      |        |   https://github.com/Oneflow-Inc/oneflow/pull/5671    |      | math_ops.py|

**Comment by doombeaker** [comment]
# 添加 docstring 的完整计划

| functional API                 | 认领人 | reviewer | 备注 |
| ------------------------------ | ------ | -------- | ---- |
| F.abs                          |        |          |      |
| F.acos                         |        |          |      |
| F.acosh                        |        |          |      |
| F.add                          |        |          |      |
| F.add_n                        |        |          |      |
| F.add_scalar                   |        |          |      |
| F.add_scalar_by_tensor         |        |          |      |
| F.argmax                       |        |          |      |
| F.argwhere                     |        |          |      |
| F.asin                         |        |          |      |
| F.asinh                        |        |          |      |
| F.atan                         |        |          |      |
| F.atanh                        |        |          |      |
| F.avg_pool_2d                  |        |          |      |
| F.batch_matmul                 |        |          |      |
| F.bias_add                     |        |          |      |
| F.broadcast_add                |        |          |      |
| F.broadcast_div                |        |          |      |
| F.broadcast_equal              |        |          |      |
| F.broadcast_greater            |        |          |      |
| F.broadcast_greater_equal      |        |          |      |
| F.broadcast_less               |        |          |      |
| F.broadcast_less_equal         |        |          |      |
| F.broadcast_like               |        |          |      |
| F.broadcast_matmul             |        |          |      |
| F.broadcast_mul                |        |          |      |
| F.broadcast_not_equal          |        |          |      |
| F.broadcast_sub                |        |          |      |
| F.cast                         |        |          |      |
| F.ceil                         |        |          |      |
| F.clip_by_scalar               |        |          |      |
| F.clip_by_scalar_max           |        |          |      |
| F.clip_by_scalar_min           |        |          |      |
| F.concat                       |        |          |      |
| F.constant                     |        |          |      |
| F.conv1d                       |        |          |      |
| F.conv2d                       |        |          |      |
| F.copy                         |        |          |      |
| F.cosh                         |        |          |      |
| F.diag                         |        |          |      |
| F.dim_gather                   |        |          |      |
| F.dim_scatter                  |        |          |      |
| F.dim_scatter_add              |        |          |      |
| F.dim_scatter_add_scalar       |        |          |      |
| F.dim_scatter_mul              |        |          |      |
| F.dim_scatter_mul_scalar       |        |          |      |
| F.dim_scatter_scalar           |        |          |      |
| F.div_scalar_by_tensor         |        |          |      |
| F.dropout                      |        |          |      |
| F.elu                          |        |          |      |
| F.erf                          |        |          |      |
| F.exp                          |        |          |      |
| F.expand                       |        |          |      |
| F.expand_dims                  |        |          |      |
| F.expm1                        |        |          |      |
| F.flatten                      |        |          |      |
| F.floor                        |        |          |      |
| F.gather                       |        |          |      |
| F.gather_nd                    |        |          |      |
| F.gelu                         |        |          |      |
| F.hardsigmoid                  |        |          |      |
| F.hardswish                    |        |          |      |
| F.hardtanh                     |        |          |      |
| F.image_flip                   |        |          |      |
| F.layer_norm                   |        |          |      |
| F.layer_norm_affine            |        |          |      |
| F.leaky_relu                   |        |          |      |
| F.lgamma                       |        |          |      |
| F.log                          |        |          |      |
| F.log1p                        |        |          |      |
| F.log_sigmoid                  |        |          |      |
| F.logical_slice                |        |          |      |
| F.logical_slice_assign         |        |          |      |
| F.matmul                       |        |          |      |
| F.max_pool_2d                  |        |          |      |
| F.maxpool_2d                   |        |          |      |
| F.maxpool_3d                   |        |          |      |
| F.mul                          |        |          |      |
| F.mul_scalar                   |        |          |      |
| F.mul_scalar_by_tensor         |        |          |      |
| F.negative                     |        |          |      |
| F.normalization                |        |          |      |
| F.ones_like                    |        |          |      |
| F.pad                          |        |          |      |
| F.pow                          |        |          |      |
| F.pow_scalar                   |        |          |      |
| F.prelu                        |        |          |      |
| F.range                        |        |          |      |
| F.reciprocal                   |        |          |      |
| F.reciprocal_no_nan            |        |          |      |
| F.reduce_mean                  |        |          |      |
| F.reduce_sum                   |        |          |      |
| F.relu                         |        |          |      |
| F.reshape                      |        |          |      |
| F.rint                         |        |          |      |
| F.round                        |        |          |      |
| F.rsqrt                        |        |          |      |
| F.scalar_add                   |        |          |      |
| F.scatterndlike                |        |          |      |
| F.sigmoid                      |        |          |      |
| F.sign                         |        |          |      |
| F.sinh                         |        |          |      |
| F.slice                        |        |          |      |
| F.slice_update                 |        |          |      |
| F.smooth_l1_loss               |        |          |      |
| F.softmax                      |        |          |      |
| F.softplus                     |        |          |      |
| F.sparse_softmax_cross_entropy |        |          |      |
| F.sqrt                         |        |          |      |
| F.square                       |        |          |      |
| F.squeeze                      |        |          |      |
| F.sub_scalar_by_tensor         |        |          |      |
| F.tanh                         |        |          |      |
| F.tensor_getitem               |        |          |      |
| F.tensor_setitem               |        |          |      |
| F.transpose                    |        |          |      |
| F.triu                         |        |          |      |
| F.upsample                     |        |          |      |
| F.upsample_bicubic_2d          |        |          |      |
| F.upsample_bilinear_2d         |        |          |      |
| F.upsample_linear_1d           |        |          |      |
| F.upsample_nearest_1d          |        |          |      |
| F.upsample_nearest_2d          |        |          |      |
| F.upsample_nearest_3d          |        |          |      |
| F.upsample_trilinear_3d        |        |          |      |
| F.where                        |        |          |      |
| F.zeros_like                   |        |          |      |

**Comment by doombeaker** [comment]
## docstring 与 doctest

https://github.com/Oneflow-Inc/OneTeam/issues/94

以上是 docstring 的介绍，doctest 呢，是 docstring 中的一种约定规范，即：如果我们按照某种约定，书写 docstring，

那么，我们可以利用 python 的标准包 `doctest`， 去提取 docstring 中的“代码”部分，达到自动测试的目的。

对于支持 docstest 的py文件，可以通过：



```
python 文件名.py
```

测试：

```
python3 /home/yaochi/oneflow/python/oneflow/nn/modules/math_ops.py -v
```

-v 会显示过程中的信息。



### 文档编译

```
cd oneflow/docs
```

安装相关的依赖：

```
pip install -U -r requirements.txt 
```

编译文档：

```
make html SPHINXOPTS="-W --keep-going"
```

以上的命令是CI中命令，如果出现警告，也是不能编译通过的。

编译完成后 `docs/build/html` 目录中包含了网页内容。



### sphinx 工具的原理是什么

sphinx 是一个 docstring 动态提取的工具，换言之，它会 `import oneflow`，然后根据 `rst` 文件中的配置，去提取 docstring。

所以，我们一定要确保，当前环境的 oneflow，与你正在修改的oneflow 版本，是一致的，否则不会生效。

```
. ~/oneflow/build/source.sh
```



## oneflow.F 名称空间

`oneflow.F` 是一个非常特殊的名称空间，它里面的方法涉及的代码，其实都是 C++ 代码。

`oneflow.F` 中的方法，都是通过代码运行时，动态加入到这个 `F` 空间的。

为了给 `oneflow.F` 动态添加 docstring，我们提供了 `add_docstr` 的方法，具体可以参考：

https://github.com/Oneflow-Inc/oneflow/pull/5449/files



```
from oneflow.python.framework.docstr.utils import add_docstr
```





## 为 oneflow.F 中的方法，添加 docstirng

有以下几点要求：

- 添加 docstring（记得以人为本，现有master中的 "\n" 是因为自动生成的，不是要求，并不推荐）
- 记得先完成本地的测试
- 记得文件管理（这个可以和我提前沟通）



测试：

```
python /home/yaochi/oneflow/python/oneflow/test/modules/test_functional_docstr.py
```



具体要求：

今天下午16:00前，需要提交各自的 PR。

demo 分工：见 https://github.com/Oneflow-Inc/oneflow/issues/5545

**Comment by GehangZhang** [comment]
等待确认的计划

| functional API            | 认领人 | reviewer | PR    | 备注 | 文件            |
| ------------------------- | ------ | -------- | ----- | ---- | --------------- |
| F.abs                     | 张舸航 |          | #5632 |      | **math_ops.py** |
| F.ceil                    | 张昌昊 |          |       |      | math_ops.py     |
| F.diag                    | 谢子鹏 |          |       |      | array_ops.py    |
| F.exp                     | 党凯   |          |       |      | math_ops.py     |
| F.floor                   | 刘圻   |          |       |      | math_ops.py     |
| F.acos                    | 张舸航 |          |       |      | math_ops.py     |
| F.acosh                   | 谢子鹏 |          |       |      | math_ops.py     |
| F.add                     | 谢子鹏 |          |       |      | math_ops.py     |
| F.add_n                   | 谢子鹏 |          |       |      | math_ops.py     |
| F.add_scalar              | 谢子鹏 |          |       |      | math_ops.py     |
| F.add_scalar_by_tensor    | 张舸航 |          |       |      | math_ops.py     |
| F.argmax                  | 谢子鹏 |          |       |      | math_ops.py     |
| F.argwhere                | 谢子鹏 |          |       |      | array_ops.py    |
| F.asin                    | 谢子鹏 |          |       |      | math_ops.py     |
| F.asinh                   | 谢子鹏 |          |       |      | math_ops.py     |
| F.atan                    | 谢子鹏 |          |       |      | math_ops.py     |
| F.atanh                   | 张舸航 |          |       |      | math_ops.py     |
| F.avg_pool_2d             | 党凯   |          |       |      | array_ops.py    |
| F.batch_matmul            | 党凯   |          |       |      | math_ops.py     |
| F.bias_add                | 党凯   |          |       |      | math_ops.py     |
| F.broadcast_add           | 张舸航 |          |       |      | math_ops.py     |
| F.broadcast_div           | 张舸航 |          |       |      | math_ops.py     |
| F.broadcast_equal         | 党凯   |          |       |      | math_ops.py     |
| F.broadcast_greater       | 张舸航 |          |       |      | math_ops.py     |
| F.broadcast_greater_equal | 党凯   |          |       |      | math_ops.py     |
| F.broadcast_less          | 张舸航 |          |       |      | math_ops.py     |
| F.broadcast_less_equal    | 张舸航 |          |       |      | math_ops.py     |
| F.broadcast_like          | 张舸航 |          |       |      | array_ops.py    |
| F.broadcast_matmul        | 党凯   |          |       |      | math_ops.py     |
| F.broadcast_mul           | 党凯   |          |       |      | math_ops.py     |
| F.broadcast_not_equal     | 张舸航 |          |       |      | math_ops.py     |
| F.broadcast_sub           | 张舸航 |          |       |      | math_ops.py     |
| F.cast                    | 党凯   |          |       |      |                 |
| F.clip_by_scalar          | 党凯   |          |       |      | array_ops.py    |
| F.clip_by_scalar_max      | 党凯   |          |       |      | array_ops.py    |
| F.clip_by_scalar_min      | 党凯   |          |       |      | array_ops.py    |



## math

  abs,

  acos,

  acosh,

  asin,

  asinh,

  atan,

  atanh,

  ceil,

  cos,

  cosh,

  exp,

  expm1,

  floor,

  rint,

  round,

  rsqrt,

  sign,

  sin,

  sinh,

  sqrt,

  square,

  tan,

  tanh,

  add,

  add_n,

  argmax,
